drivers: wifi: siwx91x: Split the driver

jerome-pouiller opened the pull request:
The WiFi driver for siwx91x start to be large and a bit confusing. This PR split in smaller pieces and clearly identify the functions shared between the sub parts.

jerome-pouiller committed:
drivers: wifi: siwx91x: Extract scan related functions

siwx91x_wifi.c starts to contains to much code. Let's simplify it by
grouping all the scan related functions in a separated file.

Signed-off-by: Jérôme Pouiller <jerome.pouiller@silabs.com>

jerome-pouiller committed:
drivers: wifi: siwx91x: Extract AP related functions

siwx91x_wifi.c starts to contains to much code. Let' simplify it by
grouping all the Access Point related functions in a separated file.

Signed-off-by: Jérôme Pouiller <jerome.pouiller@silabs.com>

jerome-pouiller committed:
drivers: wifi: siwx91x: Extract Power Save related functions

siwx91x_wifi.c starts to contains to much code. Let's simplify it by
grouping all the power-save related functions in a separated file.

Signed-off-by: Jérôme Pouiller <jerome.pouiller@silabs.com>

jerome-pouiller committed:
drivers: wifi: siwx91x: Fix siwx91x_wifi_socket.h syntax

Functions in siwx91x_wifi_socket.h are not declared static. So it create
duplicated symbols if the file is included twice.

Signed-off-by: Jérôme Pouiller <jerome.pouiller@silabs.com>

jerome-pouiller committed:
drivers: wifi: siwx91x: Extract station related functions

siwx91x_wifi.c starts to contains to much code. Let' simplify it by
grouping all the station related functions in a separated file.

Signed-off-by: Jérôme Pouiller <jerome.pouiller@silabs.com>

jerome-pouiller committed:
drivers: wifi: siwx91x: Clean up included files

To avoid confusion, we prefer to only keep header files really used by the
source file.

Signed-off-by: Jérôme Pouiller <jerome.pouiller@silabs.com>

jerome-pouiller committed:
drivers: wifi: siwx91x: Align the ways errors are managed

Unify the way the error code are tested:
  - sl_status_t is an alias to an int. The compiler won't complain if an
    int is assigned or compared to a sl_status_t. In the other hand,
    "int" is well known to manage error code. So, just use "int"
    everywhere.
  - Always name the return code "ret"
  - Unless it is meaningful, test the value of ret against 0
    (SL_STATUS_OK is just a convoluted way to say 0).

Signed-off-by: Jérôme Pouiller <jerome.pouiller@silabs.com>

jerome-pouiller committed:
drivers: wifi: siwx91x: Group ARG_UNUSED() after declarations

We want to avoid declarations after statements.

Signed-off-by: Jérôme Pouiller <jerome.pouiller@silabs.com>

jerome-pouiller committed:
drivers: wifi: siwx91x: Uniformize declaration of 'interface'

sl_wifi_interface_t interface is always initialized with
sl_wifi_get_default_interface().

Let's assign the value during the declaration everywhere.

Signed-off-by: Jérôme Pouiller <jerome.pouiller@silabs.com>

jerome-pouiller committed:
drivers: wifi: siwx91x: Uniformize error management

Usually, the exception are treated in a condition before the nominal
case:

    ret = f();
    if (ret) {
        return -EXXX;
    }
    ... nominal case ...

siwx91x_on_join_ipv4() and siwx91x_on_join_ipv6() didn't follow this
pattern.

Signed-off-by: Jérôme Pouiller <jerome.pouiller@silabs.com>

jerome-pouiller committed:
drivers: wifi: siwx91x: Do not set unused values

We avoid to assign unused values during declaration. Hence, the reader
knows that every value assigned are meaningful. It also allow compile to
raise true warning if a value is used uninitialized.

Signed-off-by: Jérôme Pouiller <jerome.pouiller@silabs.com>

ArunmaniAlagarsamy2710 reviewed (commented):
Thanks for the PR. Just one concern regarding the AP event, otherwise LGTM.

jerome-pouiller committed:
drivers: wifi: siwx91x: Prefer WIFI_MAC_ADDR_LEN

The preferred way to reference the size of the MAC address is
WIFI_MAC_ADDR_LEN.

For consistency, replace ARRAY_SIZE(mac.octet) with WIFI_MAC_ADDR_LEN.

Signed-off-by: Jérôme Pouiller <jerome.pouiller@silabs.com>

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi_sta.c:
```diff
@@ -0,0 +1,355 @@
+		break;
+	case STATE_ASSOCIATED:
+		if (FIELD_GET(SIWX91X_INTERFACE_MASK, interface) == SL_WIFI_AP_INTERFACE) {
+			wifi_mgmt_raise_ap_enable_result_event(sidev->iface,
```
Good catch. Something went wrong when I have rebased on the last version of the upstream.

jerome-pouiller committed:
drivers: wifi: siwx91x: Assign variable on declaration

When a variable is assigned once, we try to assign it during the
declaration.

Signed-off-by: Jérôme Pouiller <jerome.pouiller@silabs.com>

jerome-pouiller commented:
v2:
  - Add two cosmetics commits
  - 73a1bef was dropped during my last rebase

jerome-pouiller commented on drivers/wifi/siwx91x/siwx91x_wifi_ap.c:
```diff
@@ -0,0 +1,342 @@
+		return -EINVAL;
+	}
+
+	memcpy(mac.octet, mac_addr, ARRAY_SIZE(mac.octet));
```
:heavy_check_mark:
